### PR TITLE
add test for admin user model

### PR DESCRIPTION
**Pull request by TheTallJerry**
## Motivation and Context

Added tests for uncovered method in the AdminUser model. Used constant in AdminUser model. 
## Your Changes

**Description**: Added a new test file under spec/models for the AdminUser model, testing both the internal behavior and the return value of the tested method. In the same model, for the call to `find_or_create_by`, changed the passed in string to a predefined constant. This change doesn't affect other areas of the application. 

**Type of change** (select all that apply):

- [x] Refactoring (internal change to codebase, without changing functionality)
- [x] Test update (change that modifies or updates tests only)

## Testing

Ran RSpec with this newly added test file. Got 100% coverage for the method covered and all the exisiting tests passed. 

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the pre-commit.ci checks have passed. 
- [ ] I have verified that the CI tests have passed. 
- [ ] I have reviewed the test coverage changes reported on Coveralls.

### CODE REVIEW

**Review by david-yz-liu — COMMENTED** [review]
@TheTallJerry great start! I think there are some other cases of arguments that you should be testing as well.

**Comment by TheTallJerry** [comment]
Looks like CI failed - looking into it

File: spec/models/admin_user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+describe AdminUser do
+  it 'return type should be User' do
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:2]: Tests for a method are grouped together in a separate `describe` block (see our other test files). In this case it would be `describe '.find_or_create'`.

File: spec/models/admin_user_spec.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    testuser = AdminUser.find_or_create
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:3]: Instead of calling this in each method, pull it out into a [`let` statement](https://relishapp.com/rspec/rspec-core/docs/helper-methods/let-and-let) inside the describe block for this method.

File: spec/models/admin_user_spec.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    expect(testuser).to be_a_kind_of(User)
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:4]: In general we don't test for types, but you can keep this test case but change it to `AdminUser` to make it more specific.

File: spec/models/admin_user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  end
+  context 'no existing admin' do
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:6]: For context blocks, the text should introduce a condition with "when"/"where" and use complete phrases. For example, `'when there is no existing admin'`.

File: spec/models/admin_user_spec.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    it 'first_name and last_name should equal to the specified values' do
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:7]: Again, the labels are meant to form human phrases, with "it" referring to the method. So for example, `it 'returns a user...'`

Also: "specified value" is vague. You can instead use "default value".

Finally, split the two expect statements into separate tests.

File: spec/models/admin_user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+      testuser = AdminUser.find_or_create
+      expect(testuser.first_name).to eq('admin')
+      expect(testuser.last_name).to eq('admin')
+    end
+  end
+  context 'existing admin' do
+    it 'first_name and last_name should ot be changed for existing admin' do
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:14]: In addition to my comments above, there's a typo with "ot". And once again, make sure the sentence makes grammatical sense.

Comment (david-yz-liu) [reply spec/models/admin_user_spec.rb:14]: Also make sure to split these up into separate tests as well.

You can use a `let!(:admin_user)` in this context block to ensure an admin user is already created at each test (you can check out the documentation for the difference between `let` and `let!`).

**Review by david-yz-liu — COMMENTED** [review]
@TheTallJerry okay great, I left a more detailed review this time. In addition to the inline comments, please also add a test to check that the `api_key` attribute is not `nil` (this is the other line of code in the `find_or_create` method).

File: spec/models/admin_user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  describe '.find_or_create' do
+    let(:admin_user) { AdminUser.find_or_create }
+    it 'returns an instance of AdminUser' do
+      expect(admin_user).to be_an_instance_of(AdminUser)
+    end
+    it 'ensures the returned AdminUser api_key is not null' do
+      expect(admin_user.api_key).not_to be(nil)
+    end
+    context 'when there is no existing AdminUser' do
+      it 'will create a new AdminUser' do
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:11]: In general, use present tense for the descriptions ("creates" instead of "will create").

File: spec/models/admin_user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        expect { AdminUser.find_or_create }.to change { AdminUser.count }.by(1)
+      end
+      it 'will set first_name and last_name to the default values' do
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:14]: split this up into two separate tests

File: spec/models/admin_user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        expect(admin_user.first_name).to eq('admin')
+        expect(admin_user.last_name).to eq('admin')
+      end
+    end
+    context 'when there is existing AdminUser' do
+      let!(:existing_admin_user) do
+        AdminUser.create(user_name: AdminUser::ADMIN_USERNAME, first_name: 'jerry', last_name: 'foo')
+      end
+      let(:returned_user) { AdminUser.find_or_create }
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:23]: It's preferable to reuse the same object as before (`admin_user`). I don't think you need to worry about the variable name being less explicit, since the tests are short and will rely on the English descriptions to convey what they're doing.

File: spec/models/admin_user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+      it 'will not create a new AdminUser' do
+        expect { AdminUser.find_or_create }.not_to(change { AdminUser.count })
+      end
+      it 'will return the existing AdminUser' do
+        expect(returned_user).to eq(existing_admin_user)
+      end
+      it 'will leave first_name and last_name unchanged' do
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:30]: same comment as above about splitting

File: spec/models/admin_user_spec.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
+describe AdminUser do
+  describe '.find_or_create' do
+    let(:admin_user) { AdminUser.find_or_create }
+    it 'returns an instance of AdminUser' do
+      expect(admin_user).to be_an_instance_of(AdminUser)
+    end
+    it 'ensures the returned AdminUser api_key is not null' do
```
Comment (david-yz-liu) [inline spec/models/admin_user_spec.rb:7]: Oops I missed this, but please also change "null" to "nil".

**Review by david-yz-liu — APPROVED** [review]
